document/document_test.go

```diff
@@ -286,7 +290,7 @@ func TestNewFromMap(t *testing.T) {
 
 func TestNewFromStruct(t *testing.T) {
```

Comment:
This test is here to indicate if the `NewFromStruct` implementation is correct or not, it should not be modified for no reason. I can see from the modifications you wrote below that the new implementation is not correct anymore.
We should undo the modifications below (except the one directly related to the embed struct of course) and modify `NewFromStruct` until it passes all the tests